### PR TITLE
MPS: Changed the apply_measure algorithm

### PR DESCRIPTION
**Pull request by merav-aharoni**
### Summary
Changed the `MPS::apply_measure` method so that it does not move the qubits to sorted ordering before measuring.

### Details and comments
This PR addresses issue #1694.
 Changed the `apply_measure` algorithm so that it does not move the qubits to the sorted positions before measuring. Instead, we measure the qubits in the order of the current `MPS` structure and then sort the measurement outcomes as a post-process. Beside fixing the issue, this is expected to improve performance and enable more accurate representation of entangled states. For example, the circuit provided in issue #1694 cannot be represented accurately when the `MPS` structure is in sorted ordering, but can be represented accurately with the new `apply_measure`.
Note that this modification does not actually fix the bug. It is possible that there is still a bug somewhere in the code. This should be addressed separately.

### CODE REVIEW

**Comment by merav-aharoni** [comment]
A note for the reviewers: `AerSimulator` actually sorts the qubits to be measured and removes the qubits that are not measured. Therefore, I could have assumed this in the call to `apply_measure_internal` and we could have avoided some of the internal sorting processes. Nevertheless, I preferred not to rely on this assumption. This adds some complication to the code, but makes it more robust.

File: src/simulators/matrix_product_state/matrix_product_state_internal.cpp [hunk 1]
```diff
@@ -1573,6 +1579,36 @@ void MPS::propagate_to_neighbors_internal(uint_t min_qubit, uint_t max_qubit,
...
+             sub_ordering[next] = qubit_ordering_.order_[i];
+             next++;
+             break;
+          }
+      }
+  }
+  return sorted_qubits;
+}
+
+reg_t MPS::sort_measured_values(reg_t input_outcome, reg_t& sub_ordering) {
```
Comment (yaelbh) [inline src/simulators/matrix_product_state/matrix_product_state_internal.cpp:1598]:
```suggestion
reg_t MPS::sort_measured_values(const reg_t& input_outcome, reg_t& sub_ordering) {
```

**Review by yaelbh — APPROVED** [review]
It was tough to review it, and follow all the manipulations that all these arrays go through: `sorted_qubits, sub_ordering, outcome_vector, sorted_outcome_vector`. I did it slowly with an example, and it looks correct. It will be nice if you find a way to somehow facilitate the code reading.

**Comment by merav-aharoni** [comment]
@hhorii , Can you please review this PR? @yaelbh already reviewed and approved. I made the changes she requested.

**Review by hhorii — APPROVED** [review]
Change looks good to me.

**Pull request merged**